### PR TITLE
Explore: Add active state to ContentOutlineItemButton

### LINKED ISSUE #79808: New click state and highlight when scrolling
Highlight content outline items on scroll and click

### PR DESCRIPTION
**Pull request by harisrozajac**
This feature is part of [Content Outline v2](https://github.com/grafana/grafana/issues/78770).

- It adds an active state to ContentOutlineItemButton, so it will be highlighted when the user clicks on an item in the content outline. 
- It highlights each item on scroll based on the corresponding panel that is closest to the top but whose top part is not covered 

The default highlighted button is the first item in the content outline items array, which is always the query editor.

Demo:

https://github.com/grafana/grafana/assets/58232930/2740c6da-52cd-4683-b61b-c498c3a671dc

Fixes: #79808

Please check that:
- [x] It works as expected from a user's perspective.

### CODE REVIEW

**Comment by harisrozajac** [comment]
@grafana/explore-squad I need some feedback on how to move forward with highlighting on scroll. After doing some exploration, I've found a solution for highlighting when scrolling, but this still causes an issue when clicking directly on a button because useEffect has a `verticallScroll` dependency, which will fire useEffect as scroll is occurring, and since we have scroll set to smooth, useEffect will keep firing and cause the buttons in sidebars to highlight as they come into view before the target button. 

I've found two potential solutions:

1. Currently, I've added a timeout in the scrollIntoView to delay the useEffect from being triggered too often, but since the timeout is set to 1s this is not consistent especially if user is clicking through buttons quickly. The problem with using lower values for timeout is that if a distance between current panel and element panel is big (like there are other panels between them) then useEffect will be triggered and you'll see some flickering. 

A potential solution here could be to evaluate the distance of the target panel from the top and then determine the timing of the timeout, but it is a little hacky, and it could be hard to make it work consistently. 

2. Set scroll behavior to instant. This solves the consistency problem, but I don't know if it's great for the UX because there's no smoothness to it.

**Comment by ifrost** [comment]
> I need some feedback on how to move forward with highlighting on scroll

This may not be a very popular opinion but to be honest I like the effect of highlighting items "on the way" :D I don't see it as a UX issue and it works quite smoothly. What is more, it makes the code simpler and easier to maintain:

https://github.com/grafana/grafana/assets/745532/2cdb65f8-7699-4919-be21-07daad69a757

Out of curiosity, I checked how Google docs nav bar is working and it has exactly the same behavior, though the transition is faster:

https://github.com/grafana/grafana/assets/745532/2c241e67-3fb9-431f-9ea0-f8c3f7d43c1e

**Comment by harisrozajac** [comment]
@ifrost sounds great! 😄  i'm going to do a little bit of cleanup and get the PR out of the draft mode

File: public/app/features/explore/ContentOutline/ContentOutline.tsx [hunk 1]
```diff
@@ -77,16 +99,21 @@ export function ContentOutline({ scroller, panelId }: { scroller: HTMLElement |
...
+          {outlineItems.map((item) => {
+            return (
+              <ContentOutlineItemButton
+                key={item.id}
+                title={expanded ? item.title : undefined}
+                className={styles.buttonStyles}
+                icon={item.icon}
+                onClick={() => {
+                  scrollIntoView(item.ref, item.title);
+                }}
```
Comment (ifrost) [inline public/app/features/explore/ContentOutline/ContentOutline.tsx:111]: Nit: It doesn't change the behavior, just adds extra `{ ... }` which is not needed.

File: public/app/features/explore/ContentOutline/ContentOutline.tsx [hunk 2]
```diff
@@ -33,9 +33,12 @@ const getStyles = (theme: GrafanaTheme2) => {
 };
 
 export function ContentOutline({ scroller, panelId }: { scroller: HTMLElement | undefined; panelId: string }) {
+  const { outlineItems } = useContentOutlineContext();
   const [expanded, toggleExpanded] = useToggle(false);
+  const [activeItemId, setActiveItemId] = useState<string | undefined>(outlineItems[0]?.id);
   const styles = useStyles2((theme) => getStyles(theme));
-  const { outlineItems } = useContentOutlineContext();
+  const scrollerRef = useRef(scroller as HTMLElement);
```
Comment (ifrost) [inline public/app/features/explore/ContentOutline/ContentOutline.tsx:40]:
```suggestion
  const scrollerRef = useRef(scroller || null);
```

This should do the trick to remove the assertion and make the betterer happy ☕

**Review by ifrost — APPROVED** [review]
Left a few minor comments. Looks great 🙌

**Pull request merged**